File: CS/Student.cs
namespace StudentManagementSystem;

public class Student : Human
{
    public Student(string name, short age) : base(name, age)
    { }

    public byte GetGrade(Lecture lecture)
        => lecture.Grades[this.Name];
}

public class StudentPresident : Student
{
    public StudentPresident(string name, short age) : base(name, age)
    { }

    public void CheckAttendance(Lecture lecture, string studentName, string date, bool check)
    {
        lecture.AttendanceList.TryGetValue(studentName, out Dictionary<string, bool>? studentAttendanceList);
        if (studentAttendanceList is null)
        {
            studentAttendanceList = new();
            lecture.AttendanceList.Add(studentName, studentAttendanceList);
        }
        bool exist = studentAttendanceList.TryGetValue(date, out bool curCheck);
        if (exist)
            curCheck = check;
        else
            studentAttendanceList.Add(date, check);
    }
}


public class StudentVicePresident : Student
{
    public StudentVicePresident(string name, short age) : base(name, age)
    { }

    public void Grade(Lecture lecture, string studentName, byte grade)
    {
        bool exist = lecture.Grades.TryGetValue(studentName, out byte curGrade);
        if (exist)
            curGrade = grade;
        else
            lecture.Grades.Add(studentName, grade);
    }
}

File: Rust/src/student.rs
use crate::human::Human;
use crate::lecture::Lecture;
use crate::teacher::{IGrader, IAttendanceChecker};

use crate::{human_getters, human_new};

pub trait IStudent {
	fn get_grade(&self, lecture: &mut Lecture) -> i8;
}

macro_rules! student_function {
	{} => {
		fn get_grade(&self, lecture: &mut Lecture) -> i8 {
			match lecture.get_grades().get(&self.name) {
				Some(&grade) => grade,
				_ => -1
			}
		}
	}
}

#[derive(Debug, Clone)]
pub struct Student {
	name: String,
	age: i8,
}

impl Student {
	human_new!();
}

impl Human for Student {
	human_getters!();
}

impl IStudent for Student {
	student_function!();
}

#[derive(Debug, Clone)]
pub struct StudentPresident {
	name: String,
	age: i8,
}

impl StudentPresident {
	human_new!();
}

impl Human for StudentPresident {
	human_getters!();
}

impl IStudent for StudentPresident {
	student_function!();
}

impl IAttendanceChecker for StudentPresident {}

#[derive(Debug, Clone)]
pub struct StudentVicePresident {
	name: String,
	age: i8,
}

impl StudentVicePresident {
	human_new!();
}

impl Human for StudentVicePresident {
	human_getters!();
}

impl IStudent for StudentVicePresident {
	student_function!();
}

impl IGrader for StudentVicePresident {}
